lib/IDE/IDETypeChecking.cpp

```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
   return Scratch;
 }
 
+/// This walker will traverse the AST and report types for every variable declaration.
+class VariableTypeCollector: public SourceEntityWalker {
+private:
+  SourceManager &SM;
+  unsigned int BufferId;
+  std::vector<VariableTypeInfo> &Results;
+
+  /// This is where we print all types.
+  llvm::raw_ostream &OS;
+
+  /// Map from a printed type to the offset in OS where the type starts.
+  llvm::StringMap<uint32_t> TypeOffsets;
+
+  /// Specified by the client whether we should canonicalize types before printing.
+  const bool CanonicalType;
+
+  /// Find an exicisting offset in the type buffer, otherwise print the type.
+  std::pair<uint32_t, uint32_t> getTypeOffsets(StringRef PrintedType) {
+    auto It = TypeOffsets.find(PrintedType);
+    if (It == TypeOffsets.end()) {
+      TypeOffsets[PrintedType] = OS.tell();
+      OS << PrintedType << '\0';
+    }
+    return {TypeOffsets[PrintedType], PrintedType.size()};
+  }
+public:
+  VariableTypeCollector(SourceFile &SF,
+                        std::vector<VariableTypeInfo> &Results,
+                        bool CanonicalType,
+                        llvm::raw_ostream &OS) : SM(SF.getASTContext().SourceMgr),
+                                                 BufferId(*SF.getBufferID()),
+                                                 Results(Results),
+                                                 OS(OS),
+                                                 CanonicalType(CanonicalType) {}
+
+  bool walkToDeclPre(Decl *D, CharSourceRange Range) override {
+    if (Range.isInvalid()) {
+      return true;
+    }
+    if (auto VD = dyn_cast<VarDecl>(D)) {
+      unsigned Offset = SM.getLocOffsetInBuffer(Range.getStart(), BufferId);
+      unsigned Length = Range.getByteLength();
+      // Print the type to a temporary buffer
+      SmallString<64> Buffer;
+      {
+        llvm::raw_svector_ostream OS(Buffer);
+        auto Ty = VD->getType()->getRValueType();
+        if (CanonicalType) {
+          Ty->getCanonicalType()->print(OS);
+        } else {
+          Ty->reconstituteSugar(true)->print(OS);
+        }
+      }
+      auto Ty = getTypeOffsets(Buffer.str());
```

Comment:
Since you already have a comment 
```cpp
// Print type type to a temporary buffer
``` 

above, I think it would be good to add another one here:
```cpp
// Transfer the type to `OS` if needed and get the offsets of this string in `OS`.
```